pytorch_lightning/metrics/metric.py

```diff
@@ -284,11 +284,17 @@ def persistent(self, mode: bool = False):
         for key in self._persistent.keys():
             self._persistent[key] = mode
 
-    def state_dict(self, *args, **kwargs):
+    def state_dict(self, destination=None, prefix='', keep_vars=False):
+        destination = super().state_dict(
+            destination=destination,
+            prefix=prefix,
+            keep_vars=keep_vars
+        )
         # Register metric states to be part of the state_dict
-        state_dict = super().state_dict()
         for key in self._defaults.keys():
             if self._persistent[key]:
                 current_val = getattr(self, key)
-                state_dict.update({key: current_val})
-        return state_dict
+                if not keep_vars and torch.is_tensor(current_val):
+                    current_val = current_val.detach()
```

Comment:
should something similar be done when the state is a list?
```suggestion
                if not keep_vars:
                    if torch.is_tensor(current_val):
                        current_val = current_val.detach()
                    if isinstance(current_val, list) and all(torch.is_tensor(val) for val in current_val):
                        current_val = [val.detach() for val in current_val]
```

Reply:
The _Note_ said
https://github.com/PyTorchLightning/pytorch-lightning/blob/6ab5417fd75f9e53ca5ceb28a7493da20a67adbb/pytorch_lightning/metrics/metric.py#L39-L41
Should we just check `isinstance(current_val, list)`:
```python
if isinstance(current_val, list):
    current_val = [val.detach() for val in current_val]
```
or do
```python
if isinstance(current_val, list):
    current_val = [val.detach() if torch.is_tensor(val) else val for val in current_val]
```
?

Reply:
Lets do the second option. If an user implements a custom metric, they could store something else than tensors in the metric state (not that they should) so it is better to do an extra check at this point.

Reply:
Thank you @SkafteNicki .
I've fixed it.

Following the naming of variables in https://github.com/PyTorchLightning/pytorch-lightning/blob/4edd07c1486f62c9f0bab4181840f2ea6eb7bfe0/pytorch_lightning/metrics/metric.py#L274
I change the variable name from `val` to `cur_v`.